### PR TITLE
platform/jimfuqian/10508 Change Schema Refs to URI

### PR DESCRIPTION
**Pull request by JFU-NAVA-PBC**
This PR:

Change all the schema reference from file path to URI

This is a re-work of [PR 13067](https://github.com/CDCgov/prime-reportstream/pull/13067) to keep the source structures remain the same.

Test Steps:
1. Pass all tests

## Changes
- Going through all schema files (.yml), do the requested change: i.e. change any schema reference in path syntax to URI - e.g. - schema: classpath:/metadata/hl7_mapping/resources/Location/HD.yml
- Going through all kotlin sources, do the requested change to any schema references
- Changed gradle DSL to properly provision the schemas during build time

### Testing
- [X] Tested locally?
- [X] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Linked Issues
- Fixes #issue

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?

### CODE REVIEW

File: prime-router/build.gradle.kts [hunk 1]
```diff
@@ -549,6 +565,16 @@ tasks.register("package") {
 tasks.register("quickPackage") {
     group = rootProject.description ?: ""
     description = "Package the code and necessary files to run the Azure functions skipping unit tests and migration"
+//    sourceSets {
```
Comment (mkalish) [inline prime-router/build.gradle.kts:568]: Remove?

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt [hunk 2]
```diff
@@ -44,6 +44,29 @@ object ConfigSchemaReader : Logging {
...
+        schemaClass: Class<out ConfigSchema<out ConfigSchemaElement>>,
+    ): ConfigSchema<*> {
+        // Load a schema including any parent schemas.  Note that child schemas are loaded first and the parents last.
+//        val schemaList = when (schemaUri.scheme) {
+//            null -> fromRelative(schemaName, folder, schemaClass)
+//            else -> fromUri(schemaUri, schemaClass)
+//        }
+
+        val schemaList = when (schemaUri.scheme) {
+            "classpath" -> fromUri(schemaUri, schemaClass)
```
Comment (mkalish) [inline prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt:58]: I don't think this is right... this means we can only support cases where it's a classpath.  The goal of this is epic is to support many schemes?

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt [hunk 3]
```diff
@@ -44,6 +44,29 @@ object ConfigSchemaReader : Logging {
         return mergedSchema
     }
 
+    fun fromFile(
+        schemaUri: URI,
+        schemaClass: Class<out ConfigSchema<out ConfigSchemaElement>>,
+    ): ConfigSchema<*> {
+        // Load a schema including any parent schemas.  Note that child schemas are loaded first and the parents last.
```
Comment (mkalish) [inline prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt:51]: I'm not following why this change is required?

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt [hunk 4]
```diff
@@ -134,7 +157,8 @@ object ConfigSchemaReader : Logging {
                 readOneYamlSchema(file.inputStream(), schemaClass)
             }
             "classpath" -> {
-                val input = javaClass.classLoader.getResourceAsStream(schemaUri.path.substring(1))
+                val schemaYml = if (schemaUri.path.endsWith(".yml")) schemaUri.path else schemaUri.path + ".yml"
```
Comment (mkalish) [inline prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt:160]: Shouldn't we need this change in all the cases here?  Also, can we just put `.yml` in the schema and avoid this extra branch?

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/converter/ConverterSchemaReader.kt [hunk 5]
```diff
@@ -2,14 +2,28 @@ package gov.cdc.prime.router.fhirengine.translation.hl7.schema.converter
...
+import java.net.URI
 
 /**
  * Read a converter schema [schemaName] from a file given the root [folder].
  * @return the validated schema
  * @throws Exception if the schema is invalid or is of the wrong type
  */
 fun converterSchemaFromFile(schemaName: String, folder: String? = null): ConverterSchema {
-    val schema = ConfigSchemaReader.fromFile(schemaName, folder, schemaClass = ConverterSchema::class.java)
+    val schemaUri = getURI(folder, schemaName)
```
Comment (mkalish) [inline prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/converter/ConverterSchemaReader.kt:14]: I'm not following this change either.  What case is this handling?

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/SchemaUtils.kt [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package gov.cdc.prime.router.fhirengine.translation.hl7.utils
+
+/**
+ * helper
+ */
+fun getURI(folder: String?, schemaName: String): String {
```
Comment (mkalish) [inline prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/SchemaUtils.kt:6]: I don't think this should be required... the class path should contain all the info about the found and we shouldn't be constructing it dynamically.  I think you're editing the old code path that supports relative pathing.

File: prime-router/src/test/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReaderTests.kt [hunk 7]
```diff
@@ -89,14 +89,15 @@ class ConfigSchemaReaderTests {
     @Test
     fun `test read schema tree from file`() {
         // This is a good schema
-        val schema = ConfigSchemaReader.readSchemaTreeRelative(
-            "ORU_R01",
-            "src/test/resources/fhirengine/translation/hl7/schema/schema-read-test-01"
+        val schema = ConfigSchemaReader.readSchemaTreeUri(
```
Comment (mkalish) [inline prime-router/src/test/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReaderTests.kt:92]: This should still work... the existing code supports both relative and reading by absolute URI.

**Review by mkalish — CHANGES_REQUESTED** [review]
I think closing in here, but still some work to be done.

- I really think we should just use `git mv` to get the mappings files into the classpath, relying on that new gradle task is brittle and unnecessary
- I don't think you need any of the code changes you made.  This ticket is 100% move and update schemas

File: prime-router/build.gradle.kts [hunk 8]
```diff
@@ -537,10 +543,20 @@ tasks.azureFunctionsPackage {
     finalizedBy("copyAzureScripts")
 }
 
+// copy metadata to main resources
```
[the remark this answers is not in the document]
Comment (JFU-NAVA-PBC) [reply prime-router/build.gradle.kts:546]: This is to leverage gradle resources provision practice - copying metadata from source tree to build directory - since we switched to URI (resources) as schema references.

this is to avoid change the current source tree structures:

<local-repo-base>
             |
             |----------> metadata
             |                           |
             |                           |--------> fhir_transforms
             |                           |
             |                           |--------> hl7_mapping
             |                           |-------------> other sub dirs
             | 
             |--------- src
             |                  |
             |                  |----------> main
             |                  |                       |
             |                  |                       |--------> resources
             |                  |                       .......

 the git mv approach is OK, since the metadata hl7_mapping, fhir_transforms will be copied to build area by convention, but the un-desirable is that it break the top level metadata folders original purpose - keep the schemas etc. in one place, now hl7_mapping, fhir_transforms are moved out, RS "metadata" are scattered around in multiple locations in source tree, I think it reduces the managibility.

later on if RS adding more "metadata", where should they go? go to top level metadata dir? or go into src/main/resources/metadata?

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt (hunk 2 above)
Comment (JFU-NAVA-PBC) [reply prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt:58]: how many scheme we are going to support by this PR?

File: prime-router/build.gradle.kts (hunk 8 above)
[the remark this answers is not in the document]
Comment (mkalish) [reply prime-router/build.gradle.kts:546]: Can we not just put them in a directory called metadata?  I'm not sure anything else has to change.

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt (hunk 2 above)
Comment (mkalish) [reply prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt:58]: The code already supports all the schemes

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt (hunk 4 above)
Comment (JFU-NAVA-PBC) [reply prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt:160]: yea, here I put some heuristic tweaking, the code could be tolerant or restrict, can mandate all schema reference to yml file be explicitly ending with .yml

but even with that required in authoring schema yaml files, do we want to be tolerant (guess a suffix) or throw error e.g. can not read schema: classpath: /.../xyz ??

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/converter/ConverterSchemaReader.kt (hunk 5 above)
Comment (JFU-NAVA-PBC) [reply prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/converter/ConverterSchemaReader.kt:14]: getURI(folder, schemaName) is an adaptor func I added, it will adapt some existing local file system semantics based schema loading code so that they will work with schemas references in URI format.

I think those schema loading funcs with folder, schemaName parameters - should be deprecated since it is switched to resource URI semantics, but could be done later, for now I just want to make things working.

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/SchemaUtils.kt (hunk 6 above)
Comment (JFU-NAVA-PBC) [reply prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/SchemaUtils.kt:6]: yup, adapting old code with URI semantics....

File: prime-router/src/test/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReaderTests.kt (hunk 7 above)
Comment (JFU-NAVA-PBC) [reply prime-router/src/test/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReaderTests.kt:92]: why still keep the relative path tests since we are switching to schema reference with resource URI?

think some legacy tests either be refactored or replaced with URI based

File: prime-router/build.gradle.kts (hunk 8 above)
[the remark this answers is not in the document]
Comment (JFU-NAVA-PBC) [reply prime-router/build.gradle.kts:546]: "Can we not just put them in a directory called metadata? I'm not sure anything else has to change."

Can do that.

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt (hunk 4 above)
Comment (mkalish) [reply prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReader.kt:160]: My current thinking is that we should insist on the extension, but happy to discuss further.  Just seems clearer, since `classpath: /.../xyz` isn't actually a file and I was thinking we should make folks be explicit

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/converter/ConverterSchemaReader.kt (hunk 5 above)
Comment (mkalish) [reply prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/converter/ConverterSchemaReader.kt:14]: Agreed, though there is a clean up ticket at the end of this epic that capture that work.  The scope here was to try to just get the existing schemas going through the new code path

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/SchemaUtils.kt (hunk 6 above)
Comment (mkalish) [reply prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/SchemaUtils.kt:6]: See comment above, but the old code _should_ be able to exist along side this new URI pattern, but let me know if that isn't quite working out

File: prime-router/src/test/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReaderTests.kt (hunk 7 above)
Comment (mkalish) [reply prime-router/src/test/kotlin/fhirengine/translation/hl7/schema/ConfigSchemaReaderTests.kt:92]: Mostly just to keep the scope of the work smaller and making rollback easier if we need to an update

**Comment by arnejduranovic** [comment]
Closing since another PR solving the same ticket has been approved by codeowners and @JFU-NAVA-PBC : #13231
